[PORT] Nanoforged Katana, a new expensive Traitor Weapon.

Howluinb opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

Original PR : https://github.com/Whitesands13/Whitesands/pull/772 
Fair bit of improvements added on top, as the original was a bit unwieldy. Sprites fully from them.

Adds a new traitor special weapon : The Nanoforged Katana, for 18 TC. (not fixed) Intended as a Desword alternative.
Damage wise, balance is still being determined - I struggle to know what would actually be too strong or too weak for an expensive item like that. The main drawback should be its insane visibility - through light effects, loud sounds being played and obvious chat descriptions - it won't be an easily concealable weapon like a desword.

Ability wise :
When first buying this item, which comes in its own special sheathe, the lucky buyer needs to soak their katana in their blood so as to be able to use its special ability.  Subsequently, anyone who gets a hold of the blade can override the previous owner, so this isn't an exclusive type of deal.

When the katana crits an enemy, or kills a borg/simplemob, all needing to be sentient - it becomes "Primed". 
At this point, when inserted into its special sheathe, while being its owner, you will take a special stance.
When all of those conditions are fullfilled, clicking anywhere will make you dash forward at insane speed, slashing anyone in your path - although, you'll not be able to go through walls. Concludes with a melee attack onto whoever you land on, for free.

Comes with a free Primed activation when soaked in blood. renewable with enough time and readiness to lose blood/take damage.

New ability : When pressing a hud button, you can charge up a Blood Cleaver attack - damaging anyone in a 3X2 tile radius in the cardinal direction you clicked, with a cooldown on future usage.

Fair bit of blocking power, too. the 70 throw damage is mostly a meme - "why would you just lose your ultra expensive weapon permanently for a single attack ?". Can be removed without a prob.

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding.
Also put closed issues under this tag, if any. Format is as follows(must be lowercase):
closes #123456789
-->

It's a fun new expensive weapon to play around with, with a ton of animuh flavor that just isn't there for normal traitors. 
also, 
![image](https://user-images.githubusercontent.com/39249137/108040181-9cedae80-7002-11eb-9cf7-18a671218bf8.gif)
[_(Art by Decompressed Artist, from Decompressed: Nuke Ops)_](https://mspfa.com/?s=39295&p=1)

That above is the main reason why this exists, and why it has an edgy flavor to it. I could ask request from Author if necessary, although I did ask AsciiSquid who didn't mind said port.

## Testing Photographs and Procedure
<!--
Include any screenshots, videos, etc. of you testing your code with it successfully functioning.
Ideally testing should cover:
Intended use cases(IE: if you are making a shotgun, test it as you intend for it to be used.)
Potential edge cases(IE: try loading different ammo than you designed for into the shotgun.)
Please include the steps you went through for the testing(videos are exempt so long as we can see everything being done in frame). Said steps can also help us help you with any issues you encounter during development.
Pulls from Upstream are generally exempt from this.
-->
<details>

![image](https://cdn.discordapp.com/attachments/801432004058939422/946614926414331904/unknown.png)
![image](https://cdn.discordapp.com/attachments/842099941119229984/946873387022700574/unknown.png)
(that second image shows the sheathe in hand, but it will, or should, be placed on the belt slot.)

<summary>Screenshots&Videos</summary>



Put screenshots and Videos documenting testing and execution of intended behaviors here


</details>

For testing : Remove "victim.mind" from attack proc to get a prime on non-sentient mobs, allows for easy dashing.

TODO : 

- [x] Inflict bleeding when linking - left on the side due to not knowing what would be a good balance for that.
- [ ] Change attack/parry sound, add sounds for priming your weapon, both for you to know and for your enemies.
- [ ] Add a ton of lights effects, both when worn in belt, on hand, when primed or non primed.
- [ ] Modify the existing sprites to look slightly bigger, and brighter. Add a bunch of sprites for on mob sheathes, both for empty or full, and primed or unprimed. Make sure the sprites update properly.
- [ ] Add logging of all attacks, including linkings, and a BUNCH of descriptions + revamping the current ones. Need to be both concise, informative and cool/edgy (which are both two sides of the same coin anyways, let's be honest)
- [x]  Allow owners to reprime the blade when interacting with it again, in exchange for bleeding, damage & a long use time, intended for out of combat preparations.
- [ ] Hud icon to do a 2 by 3 wide slash that hits target for low damage but in a zone, and a cooldown.

Known bugs : 

- [ ] Monkeys can link to the blade. But a monkey going around dashing through people with a katana is really funny, so that might stay.
- [ ] Dashing allows you to go through directional windows oriented towards you, but not in any other direction. This eludes me.

## Changelog
:cl: AsciiSquid, Howluinb31153, Ivanmixo
add: A brand-new traitor weapon, the Nanoforged Katana, for 18 tcs.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Howluinb committed:
First draft

god why is this so hard

Howluinb committed:
missing dash

added back

Howluinb committed:
basics done

finally got the basics to work, i'm overjoyed

BrumalLake commented on code/game/objects/items/weebstick.dm:
```diff
@@ -0,0 +1,208 @@
+		COOLDOWN_START(src, katanalinking_cd, 5 SECONDS) // doesnt work ? no runtime/bug
+		owner = user
+		var/opposite_active_arm = BODY_ZONE_L_ARM
+		if(!owner.active_hand_index % 2)
```
what
modulo on bool

Howluinb replied:
deleted the previous post - overdumped info that was unrelated to what you were asking.

`active_hand_index` isn't a bool, you can have more than 2 hands, the modulo returns 0 or 1 depending on if it's a right or left one.

BrumalLake replied:
but you put a bang on it

Howluinb replied:
thank you - i hadn't noticed it because i was using the sword in my right constantly, and said result was what i had expected - fixed it !

Howluinb committed:
Merge remote-tracking branch 'upstream/master'

Howluinb committed:
fixed?

i hope i didn't majorly fuckup again

Howluinb committed:
cleanup

removed a lot of bad looking // comments, and removed debug things i left in by accident and didn't notice. apologies for the bad writing

R3pID commented:
MEMORIES BROKEN

DerpiestBruh commented:
THE TRUTH GOES UNSPOKEN

ike709 commented:
Imagine adding a real overpowered katana when it's already existed as a joke item for over a decade.

ToasterBan commented:
> Imagine adding a real overpowered katana when it's already existed as a joke item for over a decade.

yeah but this one lets me make references to the video game Metal Gear Rising: Revengeance by Kojima Productions and Platinum Games

FriendlyContractor commented:
This is not an 18TC item in its current state. While it has passable blocking and ability to block projectiles it only deals 26 damage on hit(which is quite bad considering existing katana deals 40 damage and 26 is even less than much cheaper esword). Im really not sure if special attack can be considered anything more than a meme considering how it deals 45 damage which would be nice if not for the fact its only usable once after you managed to murder someone without getting killed yourself

  **Now some suggestion how could this be fixed - im sure im gonna trigger someone here. Note i dont mean all of them should be implemented but rather its a list of ideas**
* Up the damage to 40. Simple as that and frankly quite boring but it does the job
* Make special attack cooler. Dealing meagerly 45 damage for attack that requires a kill to recharge is pretty underwhelming. Have it eviscerate victims spilling their internal organs on the floor! Have it slash multiple limbs off! Get creative this attack can be much more fun that what it currently is
* Goonstation has a fun katana that can hack off limbs in single hit(except head). How could would that be here!
* Have special attack not require getting a kill and simply be on cooldown. Kinda defeats the point of the weapon :/

Archanial commented on code/game/objects/items/weebstick.dm:
```diff
@@ -0,0 +1,194 @@
+/obj/item/weebstick
```
Please name it better

Howluinb replied:
done, everything renamed to "nanoforgedkatana". if that's too long, i can change it again

Archanial commented on code/game/objects/items/weebstick.dm:
```diff
@@ -0,0 +1,194 @@
+/obj/item/weebstick/proc/predash(mob/user, atom/location, params)
+	SIGNAL_HANDLER
+
+	if(!(location in view(user.client.view,user)))
```
Don't assume client and you can just check mobs view

Howluinb replied:
ah, didn't know that - modified!

Archanial commented on code/game/objects/items/weebstick.dm:
```diff
@@ -0,0 +1,194 @@
+
+/obj/item/weebstick/proc/primed_attack(atom/target, mob/living/user, obj/item/weebstick/sword)
+	var/turf/end = get_turf(user)
+	var/turf/start = get_turf(user)
```
```suggestion
	var/turf/start = end
```

Howluinb replied:
modified, now uses a diff system

Archanial reviewed (commented):
Sprites are honestly pretty bad

Howluinb commented:
> Sprites are honestly pretty bad

Yeah, they're really not gonna stay in such a state. as said in the pr, this is mostly up for seeing if people like it and knowing if it's worth to bother a spriter.
Will apply all the fixes you suggested, thank you!

---

Friendly contractor :
The damage is indeed on the lowside, same for the blocking power, but I don't think you realize how strong the katana is mid fight, when used robustly.

That special dash, allows you to go *through* people, along with doing some severe damage, and having a huge block level of 2, although with some much lower blocking power than a desword. And it primes on *softcrit*, not death.

*And it's one handed*. You can instantly sheathe it to pull out a gun, grab enemies, prevent someone from grabbing it off of you after accidentally being shoved or 'banged.  You can then instantly unsheathe it when needed, with shift+e. You can dash both to confuse enemies, or for a quick escape.
The main disadvantage of the Desword is how unwieldy it is to some degree - you basically force yourself to use only that weapon, and switching off to anything is both lengthy in actual combat, and a major risk.

I was thinking of upping the damage and armor pen, yeah. It should not be a 40 damage weapon - it should be a weapon to cut down powerful foes.

Also, a cooldown on the ability both goes against the edgy idea of "bathing your sword in blood for you to acquire its power", and makes it really hard to predict in actual combat. This is not an instant win weapon, but more of a snowball weapon. Once you managed to down an opponent in a fight and injure a bunch of others, you're that much more likely to pull off series of dashes.

--- 

Ike : I genuinely do not think this is overpowered. The dash ability is not constant, it costs *a lot*, and right now the damage values are actually really low. And it's the only actual weeb weapon that's available to traitors, without needing a tc trade.
I've also tried to make it fairly unique in its prime mechanic, and how to activate it.

VictorVow commented:
Just having this one dash ability is a bit uninteresting. Yone from LoL has a cool set of anime swordsman abilities, especially Spirit Cleave (which could be a 3 tile wide, 1-2 tile forward attack) and Soul Unbound  https://www.youtube.com/watch?v=d15O3zSLA_Q

Howluinb commented:
> Just having this one dash ability is a bit uninteresting. Yone from LoL has a cool set of anime swordsman abilities, especially Spirit Cleave (which could be a 3 tile wide, 1-2 tile forward attack) and Soul Unbound https://www.youtube.com/watch?v=d15O3zSLA_Q

Actually a good idea for the spirit cleave, yeah. Gives diversity, i'll see if i can find a way to implement it (prob a hud button). Soul unbound tho, sounds more like a wizard spell than an ability a traitor should have.

Howluinb committed:
Bugfixes

Added a CD, fixed the error lordScrubling 's pointed out & Archanial's requests, fixed a few things on the todo list which ended up revealing a lot more errors which i ended up fixing too. Lots of trial and error here, hope it's not too hacky. Thanks again for the comments, will get to the few last coding features (bleeding & seeing about repriming it for a lengthy time (out of combat thing) & cleaver attack) before looking for spriters/better sound files !

Oscari-pog commented:
THERE WILL BE BLOODSHED

Weak as fuck for 18tc and doesnt have any specific niché for that cost except being one handed.
Special ability seems like a gimmick instead of a useful one too.
Make it block all projectiles like the highlander sword to force opponents into honorable melee combat. Dunno if thats overpowered, just throwing a idea into the pot.

Howluinb commented:
> THERE WILL BE BLOODSHED
> 
> Weak as fuck for 18tc and doesnt have any specific niché for that cost except being one handed. Special ability seems like a gimmick instead of a useful one too. Make it block all projectiles like the highlander sword to force opponents into honorable melee combat. Dunno if thats overpowered, just throwing a idea into the pot.

I upped its damage a fair bit, putting it on esword level, and gave it *really* big armor pen. 
Its blocking level is not as good as a desword, but it still blocks ALL projectiles although it doesn't reflect any, and will inflict stamina damage when blocking, needing approx 7 hits of an esword to make you drop it.
The dash, from what I tested on a private server, is much stronger than you think. There's a reason I limited it to a single charge, and i'm making someone able to reprime it in advance, with enough alone time. It also deals 35 damage to a random bodypart, and most sec armor only covers the torso.
I'm also adding an ability, a blood cleave, damaging anyone in a 3x2 area in the direction the user clicks (so even backwards while running away), on a cooldown. Think a really wide slash. I'm thinking a damage of 30 right now - due to the really wide spread of damage done by it. 

With the proper tweaking of damage values, this could be a rather well balanced high cost weapon, with special abilities to make it fun and unique, along with good blocking power & general damage.

It both allows for long range fighting, and multiple targets short range fighting, while permitting easy escape or chain combos if your opponents are not careful.

It being one handed and having an easy to pull out/put in slot, is, as you pointed, an added huge benefit. It would allow for very dynamic combat, I believe.

Howluinb committed:
Cleaning up+cleaver draft

Cleaned up a few descs, a few //, fixed small things - the blade is now nearly complete mechanics wise, needing to fix only the dash through a directional window.

Added the draft for the action.dm cleaver ability - couldn't figure out how to do an overlay without doing it extremely messily, so someone said they wanted to take a look at it. don't look too hard at it.

zeskorion commented:
Ugh, another expensive gimmick murderbone item

DerpiestBruh commented:
idk man cry about it, I don't see anything wrong with a new traitor weapon

Howluinb commented:
> Ugh, another expensive gimmick murderbone item

Double esword, carp, holopara, revolver, energy chainsaw. I can't see anything else that fits.
Adding one (1) item to a list of 5 items, doesn't seem to warrant a "ughh another"

Howluinb committed:
gitignore removed

my backup files for random tests should not have been included lmfao sorry

Howluinb commented:
aaah i did a fucky wucky and accidentally removed the entire gitignore folder aaaaaaa

DatBoiTim commented on code/datums/action.dm:
```diff
@@ -552,6 +552,23 @@
+		to_chat(owner, "<span class = 'notice'>You cannot execute such a wide slash in such a small space!</span>")
+		return
+	if(!COOLDOWN_FINISHED(src, cleaver_cooldown))
+		return
```
So on discord you raised an issue of the cooldown overlay issue.

One option is to make it so when they use it before cooldown ends, they get a message directly to them saying how long is left.

IIRC there's also a way to have procs call when a cooldown ends, but I'll get back to you on that.

Howluinb replied:
oh, that option is indeed valid, but i doubt it'd be that good when you're trying to focus on battle and you have to suddenly click an icon which then makes you look at the chat, completely making you lose focus on the actual game, which is just beyond painful imo

Do indeed get back if you find a way ! and thank you !

DatBoiTim replied:
So you will need to do this

```
/datum/action/item_action/nanoforgedkatana/New(Target)
   ..()
   RegisterSignal(src, COMSIG_CD_STOP(cleaver_cooldown), .proc/cooldown_end_notif)
```

You will need to make ```/datum/action/item_action/nanoforgedkatana/proc/cooldown_end_notif()``` but that should get you a notification when cooldown ends, in lieu of the cooldown timer overlays spells have.

 This will allow you to run a proc when the cooldown ends to notify that the katana is ready.

Howluinb replied:
aaah yeah that's a good idea, and use the span class to make it appear as very big and red in chat so as to not miss it.
Good compromise, yeah! I'll test it out and report back

zeskorion commented:
> > Ugh, another expensive gimmick murderbone item
> 
> Double esword, carp, holopara, revolver, energy chainsaw. I can't see anything else that fits. Adding one (1) item to a list of 5 items, doesn't seem to warrant a "ughh another"
(also- his grace, clown car, bundles). 
these are all items generally used to murderbone, maybe with the exception of the revolver. they dont quite fit the environment this server is going for

hippie-style "spend all your points on this specific gimmick" items are terrible, especially when it's basically just "haha time to go kill all of sec, because i dont really have points to spend on anything other than the fact i've invested in killing"

zeskorion commented on code/game/objects/items/nanoforgedkatana.dm:
```diff
@@ -0,0 +1,244 @@
+	// Last ditch effort to kill someone. Obviously, terrible idea unless the situation is desperate, as it WILL embed and WILL need surgery to remove. No extra embed damage tho, so as to not make it op.
+	hitsound = 'sound/weapons/anime_slash.ogg'
+	armour_penetration = 75
+	block_level = 2
```
this is bad and infringes on desword's niche

Howluinb replied:
The armor penetration with lower damage ?
desword damage is 34 with 35 armor pen.
katana is 30 with 75 armor pen. think the damage should absolutely be lower, but the armor pen is the complete opposite. I meant this weapon as a sec/chap counter less than a "murder civilians lol"

Probably lowering the damage to like 27 ? so a 4 hit kill, lowered to 5 on sec, but can't be lowered any further?

zeskorion commented on code/game/objects/items/nanoforgedkatana.dm:
```diff
@@ -0,0 +1,244 @@
+	obj_flags = UNIQUE_RENAME
+	w_class = WEIGHT_CLASS_BULKY
+	force = 30
+	throwforce = 70
```
why the hell is its throwforce so high?

Howluinb replied:
Mostly a meme, I wanted to see if it'd be a good idea with someone more aware of balance
The main idea is : "Why would you throw your sword that costs 18 tc at an enemy with 100% chance to embed, meaning they can easily get it back without extra damage, or that you'd have to stop to commit surgery on them yourself once the battle is over?"

I was also maybe planning to add an extra sprite overlay with the sword poking out of the chest, if that ended up being seen as a good funny idea

zeskorion commented on code/game/objects/items/nanoforgedkatana.dm:
```diff
@@ -0,0 +1,244 @@
+	armour_penetration = 75
+	block_level = 2
+	block_upgrade_walk = 0 // Base desword blocking, but no possible improvements.
+	block_power = 60 //lower block power.
```
this is still insane blocking power.

Howluinb replied:
Yeah, I thought it was fairly high, but got told that "yeah well this is shit if it cant block anything !"
What do you think would be good blocking wise ? I still want a way to reach 2 block power so as to fully embrace the animuh "slashing all the bullets"

keep in mind that in most players head, the niche of desword is to reflect bullets. This doesn't reflect them, just parries them like the replica katana did.

Howluinb replied:
Also, testing wise - 60 block power means 9 hits from esword stam crits the katana.
Which means 2 shotgun blasts.

A desword can't be stamcritted by anything but a shotgun. I tried to stam crit with an esword, but you outheal *naturally* the stam damage. I'll probably lower it to 55, 50, but it's really not as strong as the 70% of desword.

zeskorion commented on code/game/objects/items/nanoforgedkatana.dm:
```diff
@@ -0,0 +1,244 @@
+	. = ..()
+	AddComponent(/datum/component/butchering, 25, 90, 5) //Not made for scalping victims, but will work nonetheless
+
+/obj/item/nanoforgedkatana/suicide_act(mob/user)
```
you should make this count as a kill for priming the katana

Howluinb replied:
Butchering or suicide ? In the first case, i'd not agree, because it's already covered by simplemobs switching from "live" to "dead" state in the attack action
in the second case, since the owner of the sword is dead, then you'd need to become its owner again ... meaning it'd reprime. However, this is a fair point on the fact that you could convince someone else to suicide then profit off of that. Adding it !

zeskorion commented on code/game/objects/items/nanoforgedkatana.dm:
```diff
@@ -0,0 +1,244 @@
+			if(victim != user)
+				var/Pdash = victim.stat
+				playsound(victim, 'sound/weapons/anime_slash.ogg', 10, TRUE)
+				victim.take_bodypart_damage(35) // High damage, but no armor penetration.
```
take_bodypart_damage doesnt check armor. this IS high armor penetration, in that it completely ignores all damage modifier. use apply_damage with a blocked var equal to a melee armor check instead.

generally, you should be using either apply_damage, for most damage sources, or adjust[type]loss, for things like DOT and such that ignore armor and species modifiers

Howluinb replied:
Aaah my apologies, i thought this counted armor, yeah it's insane then jesus.
Modifying that asap, thank you !

zeskorion replied:
thats fine, damage procs arent well documented so no one really knows which ones to use. i plan to refactor apply_damage to make it a lot easier to integrate armor into the proc

zeskorion replied:
(however i have been fixated on other shit so i aint coding jack rn)

Howluinb committed:
Restore .gitignore

Archanial commented:
Stale
